### ISSUE TITLE
.gut_editor_config.json changes at project root visible in VCS

### ISSUE THREAD

**Issue by hsandt** [bitwes/Gut#473]
.gut_editor_config.json is placed in project root so I can see it change in the VCS. I assume it's meant to be added to .gitignore?

Otherwise it could be moved to some local user settings, but I don't know how to do this in Godot. If it's not possible, Godot should add local settings / user preferences for custom tools to avoid such issues.

**Comment by bitwes** [comment]
The config is placed in `res://` because there are many settings that should be defined at the project level (directories, pre/post scripts, test prefixes, etc).  There are also a lot of user settings that can vary a lot.   The user settings could probably be broke out in the future.

Right now, there is no way to configure where the `gut_editor_config.json` file resides.  This could also be a good addition.  You can do this withe .gutconfig.json used for command line and VSCode plugin, but not for the editor.

It would also be nice to unify the the two configs.  This was not done because the settings from the panel need to be saved before each run, and you wouldn't want them clashing with the command line parameters (which script, test, inner class to be run).

**Comment by hsandt** [comment]
OK, so for now ignoring it seems the right call. I suppose this could be added somewhere in the Getting Started doc (esp. for beginners).

**Comment by hsandt** [comment]
After more testing I see that only `unit_test_name` really bothers me in the VCS: it will change when I just test another function. Unlike test `dirs` this is likely to change often and between developers, causing merge conflicts.

**Comment by bitwes** [comment]
I have an idea, that I think will work.  

* Add a button to load a config into the editor (from anywhere on your computer).
* Add a button to "save as" the editor's config (to anywhere on your computer).
  * This will never include the settings for "test name", "script name", or "inner class name".
* Editor config will ALWAYS auto-save to `user://GUT/editor_settings.json` or some similarly named place.
* When tests are run through the editor, GUT will load settings from the auto-saved file.

With this approach, multiple users will not fight over the same config file while running tests through the editor.  This also allows for a common config that is easy to load.  In addition, this saves a lot of hassle in trying to resolve values in multiple config files.

I also think shortcuts should be turned into Editor preferences instead of a file.

Thoughts?

**Comment by bitwes** [comment]
Well, this fix got a little out of hand.  But it was an interesting problem to solve.  Try out the changes in `main`.

**Comment by jamie-pate** [comment]
Now I get this in CI :(

17:53:15  Godot Engine v4.2.1.stable.official.b09f793f5 - https://godotengine.org/
17:53:19  ERROR:  Config File "user://gut_temp_directory/gut_editor_config.json" does not exist.

**Comment by hsandt** [comment]
Thanks! I was using that new version without realizing it on my new project and indeed, I didn't see any VCS diff change. I just checked the user://gut_temp_directory folder as well as editor settings in [Editor data paths](https://docs.godotengine.org/en/latest/tutorials/io/data_paths.html#editor-data-paths) and could see the config there indeed.

Then I tried to save the config, it works, although prefilling the format with .json (instead of All files with no extension) and filename with something like `gut_editor_config.json` (as content is almost the same as the actual homonymous file, except for `tests` and `unit_test_name` which are I guess too volatile) would help user get decent default values.

The downside now is that every user must remember to redefine the test config, but at least with the Save As button I can save in res://gut_editor_config_to_load.json and commit that and let other devs Load it after they clone the repo.

If it turns out that it's cumbersome to devs and they prefer keeping their current config versioned because everyone uses the same (esp. `configured_dirs`, `dirs`, `include_subdirs`, `prefix` as you mentioned in your first reply), I'm okay with that too, it's just the `unit_test_name` (apparently now renamed `selected`) that should be put elsewhere as it's really the volatile value.

So eventually if `selected` is put outside config, we can get the best of both worlds as we'd still be able to auto-load the config saved at a specific location and that would be committed and therefore updated on all devs' computers on next pull.

Maybe we can have a look at what devs say about this, decide what to do and then update the docs accordingly? (adding instructions about Save/Load)

Since my original issue is fixed I will close this issue. Then if devs want to restore some auto-load config functionality we can open new issues for them gradually.

P.S.: And cosmetics options like gut/output_font_name are good in editor settings I think, since individual devs may have different preferences for the area shown in their own editors (but the ingame "font_name": "CourierPrime" makes sense to be in main config since in pixel art games using viewport stretch mode for instance, you'll want a pixel art font for everyone or it may look blurry).
